providers/implementations/kem/hybrid_kem.c

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "prov/hybrid_pkey.h"
+#include "hybrid_kem_local.h"
+
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/core_names.h>
+#include <openssl/proverr.h>
+#include "crypto/evp.h"
+#include "prov/providercommon.h"
+#include "prov/provider_ctx.h"
+#include "internal/e_os.h"
+#include "internal/refcount.h"
+
+HYBRID_PKEY_CTX *ossl_hybrid_kem_newctx(void *provctx,
+                                        const HYBRID_ALG_INFO *info)
+{
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    /*
+     * Nothing really to do here because the PKEY_CTXs cannot be
+     * allocated until the init call because we don't have the associated
+     * PKEY that it's being build against.
+     */
+    return ossl_hybrid_pkey_ctx_alloc(PROV_LIBCTX_OF(provctx), info);
+}
+
+void *ossl_hybrid_kem_dupctx(void *vctx)
+{
+    HYBRID_PKEY_CTX *src = (HYBRID_PKEY_CTX *)vctx;
+    HYBRID_PKEY_CTX *dest = OPENSSL_zalloc(sizeof(*src));
+    unsigned int i;
+
+    if (dest == NULL)
+        return NULL;
+
+    dest->libctx = src->libctx;
+    dest->info = src->info;
+
+    if (src->propq != NULL
+            && (dest->propq = OPENSSL_strdup(src->propq)) == NULL)
+        goto err;
+
+    for (i = 0; i < src->info->num_algs; i++)
+        if (src->ctxs[i] != NULL
+                && (dest->ctxs[i] = EVP_PKEY_CTX_dup(src->ctxs[i])) == NULL)
+            goto err;
+    return dest;
+ err:
+    ossl_hybrid_pkey_ctx_free(dest);
+    return NULL;
+}
+
+int ossl_hybrid_kem_get_ctx_params(void *vctx, OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_get_ctx_params(ctx, params);
+}
+
+int ossl_hybrid_kem_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+
+    return ossl_hybrid_set_ctx_params(ctx, params);
+}
+
+static int hybrid_kem_init(HYBRID_PKEY_CTX *ctx, HYBRID_PKEY *auth,
+                           const OSSL_PARAM params[],
+                           int (*initf)(EVP_PKEY_CTX *ctx,
```

Comment:
typedef this

Reply:
Why???  The only functions passed in are existing and are drawn from a set of exactly two.
To typedef them properly, the typedef would need to be made public and the two functions declared using it.

It doesn't make a lot of sense IMO.